// super-djs/index.js
/*
    ---------------------------------------------------------------------------------------------
                                   ⚠️ STOP SCROLLING DOWN ⚠️
                                    ⚠️ READ THIS MESSAGE ⚠️

    This package is made by the developer T.F.A#7524, if you are trying to copy this package
    and publish another one with another name but same code files will ends in a copyright
    warning. Please avoid from stealing people's projects and make one for yourself by using
    your mind. This package is not a copy of any other similar packages like simply-djs. This
    was coded by hands from the developer T.F.A.

    Thank you for understanding,
    T.F.A.
    ---------------------------------------------------------------------------------------------
*/

const { name } = require('./src/config/pkg.json');

const processVersion = process.version.replace('v', '').replace('.', '').replace('.', ''); // From "v18.0.0" for an example, it will be turned to: 1800.

if (processVersion < 1690) {
    console.error(`[${name}: NODEJS_ERROR] Node.js version should be installed above or equal to v16.9.0`)
    process.exit(1);
};

const _CheckUpdates = require('./src/update/check');

_CheckUpdates();

module.exports = {
    EightBall: require('./src/modules/class/8ball'),
    GhostPingDetector: require('./src/modules/class/ghostping'),
    Wait: require('./src/modules/func/wait'),
    RandomString: require('./src/modules/func/randomString'),
    SearchNpm: require('./src/modules/class/npm'),
    MemeGenerator: require('./src/modules/class/meme-generator'),
    Calculator: require('./src/modules/class/calculator'),
    ButtonsPaginator: require('./src/modules/class/buttons-paginator'),
    MenuPaginator: require('./src/modules/class/menu-paginator'),
    isDiscordInvite: require('./src/modules/func/isDiscordInvite'),
    TextFileGenerator: require('./src/modules/class/textFileGenerator'),
    connectMongoDB: require('./src/modules/func/connectMongoDB'),
    NoteCommand: require('./src/modules/class/noteCommand'),
    packageVersion: require('./src/update/version'),
    colourText: require('./src/modules/func/colourText'),
    RPS: require('./src/modules/class/rps')
};
