// super-djs/src/modules/class/menu-paginator.js
const { name } = require('../../config/pkg.json');

try {
    require('discord.js');
} catch (err) {
    throw new Error(`[${name}: NODEJS_PACKAGE_ERROR] discord.js package is not installed, run in the terminal this command: npm i discord.js@latest`);
};

try {
    require('ms');
} catch (err) {
    throw new Error(`[${name}: NODEJS_PACKAGE_ERROR] ms package is not installed, run in the terminal this command: npm i ms`);
};

const { ActionRowBuilder, SelectMenuBuilder, ComponentType, EmbedBuilder } = require('discord.js');
const ms = require('ms');

module.exports = class MenuPaginator {
    constructor(embeds) {
        if (!embeds) throw new Error(`[${name}: MISSING_PARAMETER] The parameter 'embeds' is required to provide.`);

        if (!Array.isArray(embeds)) throw new TypeError(`[${name}: MISSING_PARAMETER] The parameter 'embeds' is not an Array.`);

        if (embeds.length <= 1) throw new Error(`[${name}: PARAMETER_ERROR] You need to provide at least two embeds.`);

        this.embeds = embeds;
    };

    configure(options =
        {
            discordApi: {
                returnEphemeral: false,
                components: {
                    placeHolder: "Click here to select",
                    expirationTime: 600000
                }
            }
        }
    ) {
        this.options = options;
        return this;
    };

    setTimeoutEmbed(embed, options = {
        showComponents: false
    }) {
        this.endEmbed = embed;
        this.showComponents = options.showComponents;
        return this;
    };

    async run(interaction) {
        if (!interaction) throw new Error(`[${name}: MISSING_PARAMETER] The parameter 'interaction' is required to provide.`);

        if (typeof this.options === "undefined") throw new Error(`[${name}: MISSING_FUNCTION] Use the function '.configure()' before '.run()' function. This function will set all the default configuration for the builder.`);

        let menu = new SelectMenuBuilder()
            .setCustomId('super_djs_select_menu')
            .setPlaceholder(this.options?.discordApi?.components?.placeHolder.toString() || "Nothing selected")
            .addOptions(
                this.embeds.map((w, i) => {
                    return {
                        label: w.data.title ? w.data.title.toString() : "Unknown title.",
                        value: i.toString()
                    }
                })
            );

        try {
            await interaction.reply({
                embeds: [
                    this.embeds[0]
                ],
                components: [
                    new ActionRowBuilder()
                        .addComponents(
                            menu
                        )
                ],
                ephemeral: this.options?.discordApi?.returnEphemeral ? true : false
            });

            const collector = interaction.channel.createMessageComponentCollector({
                filter: i => i.user.id === interaction.user.id,
                type: ComponentType.SelectMenu,
                time: this.options?.discordApi?.components?.expirationTime
            });

            collector.on('collect', async (i) => {
                if (i.customId === "super_djs_select_menu") {
                    return i.update({
                        embeds: [
                            this.embeds[i.values[0] || 0]
                        ]
                    }).catch(() => { });
                } else return;
            });

            collector.on('end', async () => {
                if (collector._endReason === "time") {
                    if (this.showComponents === true) {
                        await interaction.editReply({
                            embeds: [
                                this.endEmbed || new EmbedBuilder()
                                    .setTitle('Timed out!')
                                    .setDescription(`Your menu has been expired in **${ms(this.options?.discordApi?.components?.expirationTime, { long: true })}**.`)
                                    .setFooter({
                                        text: `By: ${name} Development.`
                                    })
                                    .setColor('Red')
                            ],
                            components: [
                                new ActionRowBuilder()
                                    .addComponents(
                                        menu.setDisabled(true)
                                    )
                            ]
                        }).catch(() => { });
                    } else {
                        await interaction.editReply({
                            embeds: [
                                this.endEmbed || new EmbedBuilder()
                                    .setTitle('Timed out!')
                                    .setDescription(`Your menu has been expired in **${ms(this.options?.discordApi?.components?.expirationTime, { long: true })}**.`)
                                    .setFooter({
                                        text: `By: ${name} Development.`
                                    })
                                    .setColor('Red')
                            ],
                            components: []
                        }).catch(() => { });
                    };
                } else return;
            });
        } catch (err) {
            throw new Error(`[${name}: ERROR] An error has been occured.\n> ` + err);
        };
    };
};
