// fakeyou-api/index.js
const p = require("phin");
const { randomUUID } = require("crypto");
const format = require('format-duration')

async function speak(voice, text, interaction) {
    let jobRequest = await p({
        url: "https://api.fakeyou.com/tts/inference",
        method: "POST",
        parse: "json",
        data: {
            uuid_idempotency_token: randomUUID(),
            tts_model_token: voice,
            inference_text: text,
        },
    });

    if (!jobRequest.body.success) {
        await interaction.editReply({ content: `❌ ¡Estás enviando demasiadas solicitudes! Reduzca la velocidad un poco.` });
        return "ERROR: 0";
    }
    return await new Promise((resolve) => {
        let count = 0;
        let interval = setInterval(async () => {
            let res = await p({
                url:
                    "https://api.fakeyou.com/tts/job/" +
                    jobRequest.body.inference_job_token,
                method: "GET",
                parse: "json",
            });

            count = count + 1 * 1000;
            if (res.body.state.status === 'pending') await interaction.editReply({ content: `En cola... (${format(count)}s)` });
            if (res.body.state.status === 'started') await interaction.editReply({ content: `Procesando...`, });
            if (count >= 180000) {
                clearInterval(interval);
                await interaction.editReply({ content: `❌ Tiempo de espera agotado, intentalo de nuevo mas tarde` });
                setTimeout(() => interaction.deleteReply(), 10000);
            }
            
            if (!res.body.success) {
                clearInterval(interval);
                resolve("ERROR: 1");
            } else if (res.body.state.status === "complete_success") {
                clearInterval(interval);
                resolve(
                    "https://storage.googleapis.com/vocodes-public" +
                        res.body.state.maybe_public_bucket_wav_audio_path
                );
            }
        }, 1000);
    });
}
exports.speak = speak;
